File: mobile/ios/BoxVistaIOS/NetworkManager.swift
//
//  NetworkManager.swift
//  BoxVistaIOS
//
//  Created by Pol Linger on 21/6/25.
//

import Foundation

enum APIError: Error, LocalizedError {
    case invalidURL
    case invalidResponse
    case decodingError
    case unauthorized(String?)
    case serverError(Int, String?)

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "URL inválida"
        case .invalidResponse:
            return "Respuesta inválida del servidor"
        case .decodingError:
            return "No se pudo procesar la respuesta del servidor"
        case .unauthorized(let message):
            return message ?? "Sesión no iniciada o caducada"
        case .serverError(let code, let message):
            return message ?? "Error del servidor (código \(code))"
        }
    }
}

private struct AuthRequest: Encodable {
    let email: String
    let password: String
}

private struct AuthResponse: Decodable {
    let token: String
}

private struct APIErrorResponse: Decodable {
    let message: String?
}

class NetworkManager {
    static let shared = NetworkManager()
    private let baseURL = URL(string: "http://localhost:3000/api")!
    private let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .convertFromSnakeCase
        return d
    }()
    private let encoder: JSONEncoder = {
        let e = JSONEncoder()
        e.keyEncodingStrategy = .convertToSnakeCase
        return e
    }()

    // MARK: - Auth

    func login(email: String,
               password: String,
               completion: @escaping (Result<String, APIError>) -> Void) {
        authenticate(path: "auth/login", email: email, password: password, completion: completion)
    }

    func register(email: String,
                  password: String,
                  completion: @escaping (Result<String, APIError>) -> Void) {
        authenticate(path: "auth/register", email: email, password: password, completion: completion)
    }

    func login(email: String, password: String) async throws -> String {
        try await withCheckedThrowingContinuation { continuation in
            login(email: email, password: password) { result in
                switch result {
                case .success(let token):
                    continuation.resume(returning: token)
                case .failure(let error):
                    continuation.resume(throwing: error)
                }
            }
        }
    }

    func register(email: String, password: String) async throws -> String {
        try await withCheckedThrowingContinuation { continuation in
            register(email: email, password: password) { result in
                switch result {
                case .success(let token):
                    continuation.resume(returning: token)
                case .failure(let error):
                    continuation.resume(throwing: error)
                }
            }
        }
    }

    private func authenticate(path: String,
                              email: String,
                              password: String,
                              completion: @escaping (Result<String, APIError>) -> Void) {
        let url = baseURL.appendingPathComponent(path)
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")

        do {
            req.httpBody = try encoder.encode(AuthRequest(email: email, password: password))
        } catch {
            return completion(.failure(.decodingError))
        }

        URLSession.shared.dataTask(with: req) { data, resp, err in
            if err != nil {
                return completion(.failure(.invalidResponse))
            }
            if let error = self.apiError(from: data, response: resp) {
                return completion(.failure(error))
            }
            guard let data = data else {
                return completion(.failure(.invalidResponse))
            }
            do {
                let authResponse = try self.decoder.decode(AuthResponse.self, from: data)
                AuthTokenStore.shared.token = authResponse.token
                completion(.success(authResponse.token))
            } catch {
                completion(.failure(.decodingError))
            }
        }.resume()
    }

    // MARK: - Boxes CRUD

    /// Fetch all boxes using DTO mapping
    func fetchBoxes(completion: @escaping (Result<[Box], APIError>) -> Void) {
        let url = baseURL.appendingPathComponent("boxes")
        var req = URLRequest(url: url)
        addAuthHeader(to: &req)

        URLSession.shared.dataTask(with: req) { data, resp, err in
            if err != nil {
                return completion(.failure(.invalidResponse))
            }
            if let error = self.apiError(from: data, response: resp) {
                return completion(.failure(error))
            }
            guard let data = data else {
                return completion(.failure(.invalidResponse))
            }
            do {
                // Decode array of DTOs and map to domain
                let dtos = try self.decoder.decode([BoxDTO].self, from: data)
                let boxes = dtos.map { $0.toBox() }
                completion(.success(boxes))
            } catch {
                completion(.failure(.decodingError))
            }
        }.resume()
    }

    /// Fetch a specific box by id
    func fetchBox(id: Int64, completion: @escaping (Result<Box, APIError>) -> Void) {
        let url = baseURL.appendingPathComponent("boxes/\(id)")
        var req = URLRequest(url: url)
        addAuthHeader(to: &req)

        URLSession.shared.dataTask(with: req) { data, resp, err in
            if err != nil {
                return completion(.failure(.invalidResponse))
            }
            if let error = self.apiError(from: data, response: resp) {
                return completion(.failure(error))
            }
            guard let data = data else {
                return completion(.failure(.invalidResponse))
            }
            do {
                let dto = try self.decoder.decode(BoxDTO.self, from: data)
                completion(.success(dto.toBox()))
            } catch {
                completion(.failure(.decodingError))
            }
        }.resume()
    }

    /// Create a new box
    func createBox(name: String,
                   description: String,
                   objects: [ObjectItem],
                   completion: @escaping (Result<Box, APIError>) -> Void) {
        let url = baseURL.appendingPathComponent("boxes")
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        addAuthHeader(to: &req)

        // Build request payload from domain objects
        let payload = [
            "name": name,
            "description": description,
            "objetos": objects.map { [
                "nombre": $0.nombre,
                "state": $0.state
            ] }
        ] as [String : Any]

        do {
            req.httpBody = try JSONSerialization.data(withJSONObject: payload, options: [])
        } catch {
            return completion(.failure(.decodingError))
        }

        URLSession.shared.dataTask(with: req) { data, resp, err in
            if err != nil {
                return completion(.failure(.invalidResponse))
            }
            if let error = self.apiError(from: data, response: resp) {
                return completion(.failure(error))
            }
            guard let data = data else {
                return completion(.failure(.invalidResponse))
            }
            do {
                let dto = try self.decoder.decode(BoxDTO.self, from: data)
                completion(.success(dto.toBox()))
            } catch {
                completion(.failure(.decodingError))
            }
        }.resume()
    }

    /// Update an existing box
    func updateBox(_ box: Box,
                   completion: @escaping (Result<Box, APIError>) -> Void) {
        let url = baseURL.appendingPathComponent("boxes/\(box.id)")
        var req = URLRequest(url: url)
        req.httpMethod = "PUT"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        addAuthHeader(to: &req)

        // Build payload similarly to create
        let payload = [
            "name": box.name,
            "description": box.description,
            "objetos": box.objects.map { [
                "nombre": $0.nombre,
                "state": $0.state
            ] }
        ] as [String : Any]

        do {
            req.httpBody = try JSONSerialization.data(withJSONObject: payload, options: [])
        } catch {
            return completion(.failure(.decodingError))
        }

        URLSession.shared.dataTask(with: req) { data, resp, err in
            if err != nil {
                return completion(.failure(.invalidResponse))
            }
            if let error = self.apiError(from: data, response: resp) {
                return completion(.failure(error))
            }
            guard let data = data else {
                return completion(.failure(.invalidResponse))
            }
            do {
                let dto = try self.decoder.decode(BoxDTO.self, from: data)
                completion(.success(dto.toBox()))
            } catch {
                completion(.failure(.decodingError))
            }
        }.resume()
    }

    /// Delete a box
    func deleteBox(id: Int64, completion: @escaping (Result<Void, APIError>) -> Void) {
        let url = baseURL.appendingPathComponent("boxes/\(id)")
        var req = URLRequest(url: url)
        req.httpMethod = "DELETE"
        addAuthHeader(to: &req)
        URLSession.shared.dataTask(with: req) { data, resp, err in
            if err != nil {
                return completion(.failure(.invalidResponse))
            }
            if let error = self.apiError(from: data, response: resp) {
                return completion(.failure(error))
            }
            completion(.success(()))
        }.resume()
    }

    // MARK: - ObjectItems CRUD

    /// Fetch all objects for a box
    func fetchObjects(for boxId: Int64,
                      completion: @escaping (Result<[ObjectItem], APIError>) -> Void) {
        let url = baseURL.appendingPathComponent("boxes/\(boxId)/objects")
        var req = URLRequest(url: url)
        addAuthHeader(to: &req)

        URLSession.shared.dataTask(with: req) { data, resp, err in
            if err != nil {
                return completion(.failure(.invalidResponse))
            }
            if let error = self.apiError(from: data, response: resp) {
                return completion(.failure(error))
            }
            guard let data = data else {
                return completion(.failure(.invalidResponse))
            }
            do {
                let dtos = try self.decoder.decode([ObjectDTO].self, from: data)
                let items = dtos.map { $0.toObjectItem() }
                completion(.success(items))
            } catch {
                completion(.failure(.decodingError))
            }
        }.resume()
    }

    /// Create a new object in a box
    func createObject(_ obj: ObjectItem,
                      in boxId: Int64,
                      completion: @escaping (Result<ObjectItem, APIError>) -> Void) {
        let url = baseURL.appendingPathComponent("boxes/\(boxId)/objects")
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        addAuthHeader(to: &req)

        let payload = [
            "nombre": obj.nombre,
            "state": obj.state
        ] as [String : Any]

        do {
            req.httpBody = try JSONSerialization.data(withJSONObject: payload, options: [])
        } catch {
            return completion(.failure(.decodingError))
        }

        URLSession.shared.dataTask(with: req) { data, resp, err in
            if err != nil {
                return completion(.failure(.invalidResponse))
            }
            if let error = self.apiError(from: data, response: resp) {
                return completion(.failure(error))
            }
            guard let data = data else {
                return completion(.failure(.invalidResponse))
            }
            do {
                let dto = try self.decoder.decode(ObjectDTO.self, from: data)
                completion(.success(dto.toObjectItem()))
            } catch {
                completion(.failure(.decodingError))
            }
        }.resume()
    }

    /// Update an object in a box
    func updateObject(_ obj: ObjectItem,
                      in boxId: Int64,
                      completion: @escaping (Result<ObjectItem, APIError>) -> Void) {
        let url = baseURL.appendingPathComponent("boxes/\(boxId)/objects/\(obj.id)")
        var req = URLRequest(url: url)
        req.httpMethod = "PUT"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        addAuthHeader(to: &req)

        let payload = [
            "nombre": obj.nombre,
            "state": obj.state
        ] as [String : Any]

        do {
            req.httpBody = try JSONSerialization.data(withJSONObject: payload, options: [])
        } catch {
            return completion(.failure(.decodingError))
        }

        URLSession.shared.dataTask(with: req) { data, resp, err in
            if err != nil {
                return completion(.failure(.invalidResponse))
            }
            if let error = self.apiError(from: data, response: resp) {
                return completion(.failure(error))
            }
            guard let data = data else {
                return completion(.failure(.invalidResponse))
            }
            do {
                let dto = try self.decoder.decode(ObjectDTO.self, from: data)
                completion(.success(dto.toObjectItem()))
            } catch {
                completion(.failure(.decodingError))
            }
        }.resume()
    }

    /// Delete an object from a box
    func deleteObject(id: Int64,
                      in boxId: Int64,
                      completion: @escaping (Result<Void, APIError>) -> Void) {
        let url = baseURL.appendingPathComponent("boxes/\(boxId)/objects/\(id)")
        var req = URLRequest(url: url)
        req.httpMethod = "DELETE"
        addAuthHeader(to: &req)
        URLSession.shared.dataTask(with: req) { data, resp, err in
            if err != nil {
                return completion(.failure(.invalidResponse))
            }
            if let error = self.apiError(from: data, response: resp) {
                return completion(.failure(error))
            }
            completion(.success(()))
        }.resume()
    }

    private func addAuthHeader(to request: inout URLRequest) {
        guard let token = AuthTokenStore.shared.token, !token.isEmpty else {
            return
        }
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
    }

    private func apiError(from data: Data?, response: URLResponse?) -> APIError? {
        guard let code = (response as? HTTPURLResponse)?.statusCode, code >= 400 else {
            return nil
        }

        let message = responseMessage(from: data)
        if code == 401 {
            return .unauthorized(message)
        }
        return .serverError(code, message)
    }

    private func responseMessage(from data: Data?) -> String? {
        guard let data, !data.isEmpty else { return nil }

        if let errorResponse = try? decoder.decode(APIErrorResponse.self, from: data),
           let message = errorResponse.message,
           !message.isEmpty {
            return message
        }

        return String(data: data, encoding: .utf8)
    }
}

File: mobile/prueba/NetworkManager.kt
package com.hirlu.boxvista

import com.hirlu.boxvista.Models.Box
import com.hirlu.boxvista.Models.BoxDTO
import com.hirlu.boxvista.Models.ObjectItem
import com.hirlu.boxvista.Models.ObjectItemDTO
import kotlinx.serialization.json.Json
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import java.io.IOException

sealed class APIError : Exception() {
    object InvalidURL : APIError()
    object InvalidResponse : APIError()
    object DecodingError : APIError()
    class ServerError(val code: Int) : APIError()
}

class NetworkManager private constructor() {
    private val baseURL = "http://localhost:3000/api"
    private val client = OkHttpClient()
    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
        prettyPrint = true
    }

    companion object {
        val shared = NetworkManager()
    }

    fun fetchBoxes(callback: (Result<List<Box>>) -> Unit) {
        val request = Request.Builder()
            .url("$baseURL/boxes")
            .build()

        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                callback(Result.failure(APIError.InvalidResponse))
            }

            override fun onResponse(call: Call, response: Response) {
                if (!response.isSuccessful) {
                    callback(Result.failure(APIError.ServerError(response.code)))
                    return
                }

                val body = response.body?.string() ?: return callback(Result.failure(APIError.InvalidResponse))
                try {
                    val dtos = json.decodeFromString<List<BoxDTO>>(body)
                    val boxes = dtos.map { it.toBox() }
                    callback(Result.success(boxes))
                } catch (e: Exception) {
                    callback(Result.failure(APIError.DecodingError))
                }
            }
        })
    }

    fun fetchBox(id: Long, callback: (Result<Box>) -> Unit) {
        val request = Request.Builder()
            .url("$baseURL/boxes/$id")
            .build()

        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                callback(Result.failure(APIError.InvalidResponse))
            }

            override fun onResponse(call: Call, response: Response) {
                if (!response.isSuccessful) {
                    callback(Result.failure(APIError.ServerError(response.code)))
                    return
                }

                val body = response.body?.string() ?: return callback(Result.failure(APIError.InvalidResponse))
                try {
                    val dto = json.decodeFromString<BoxDTO>(body)
                    callback(Result.success(dto.toBox()))
                } catch (e: Exception) {
                    callback(Result.failure(APIError.DecodingError))
                }
            }
        })
    }

    fun createBox(name: String, description: String, objects: List<ObjectItem>, callback: (Result<Box>) -> Unit) {
        val payload = mapOf(
            "name" to name,
            "description" to description,
            "objetos" to objects.map { mapOf("nombre" to it.nombre, "state" to it.state) }
        )
        val requestBody = json.encodeToString(payload).toRequestBody("application/json".toMediaType())
        val request = Request.Builder()
            .url("$baseURL/boxes")
            .post(requestBody)
            .build()
        client.newCall(request).enqueue(jsonBoxCallback(callback))
    }

    fun updateBox(box: Box, callback: (Result<Box>) -> Unit) {
        val payload = mapOf(
            "name" to box.name,
            "description" to box.description,
            "objetos" to box.objects.map { mapOf("nombre" to it.nombre, "state" to it.state) }
        )
        val requestBody = json.encodeToString(payload).toRequestBody("application/json".toMediaType())
        val request = Request.Builder()
            .url("$baseURL/boxes/${box.id}")
            .put(requestBody)
            .build()
        client.newCall(request).enqueue(jsonBoxCallback(callback))
    }

    fun deleteBox(id: Long, callback: (Result<Unit>) -> Unit) {
        val request = Request.Builder()
            .url("$baseURL/boxes/$id")
            .delete()
            .build()
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                callback(Result.failure(APIError.InvalidResponse))
            }
            override fun onResponse(call: Call, response: Response) {
                if (!response.isSuccessful) {
                    callback(Result.failure(APIError.ServerError(response.code)))
                    return
                }
                callback(Result.success(Unit))
            }
        })
    }

    fun fetchObjects(boxId: Long, callback: (Result<List<ObjectItem>>) -> Unit) {
        val request = Request.Builder()
            .url("$baseURL/boxes/$boxId/objects")
            .build()
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                callback(Result.failure(APIError.InvalidResponse))
            }
            override fun onResponse(call: Call, response: Response) {
                if (!response.isSuccessful) {
                    callback(Result.failure(APIError.ServerError(response.code)))
                    return
                }
                val body = response.body?.string() ?: return callback(Result.failure(APIError.InvalidResponse))
                try {
                    val dtos = json.decodeFromString<List<ObjectItemDTO>>(body)
                    val objs = dtos.map { it.toObject() }
                    callback(Result.success(objs))
                } catch (e: Exception) {
                    callback(Result.failure(APIError.DecodingError))
                }
            }
        })
    }

    fun createObject(obj: ObjectItem, boxId: Long, callback: (Result<ObjectItem>) -> Unit) {
        val payload = mapOf("nombre" to obj.nombre, "state" to obj.state)
        val requestBody = json.encodeToString(payload).toRequestBody("application/json".toMediaType())
        val request = Request.Builder()
            .url("$baseURL/boxes/$boxId/objects")
            .post(requestBody)
            .build()
        client.newCall(request).enqueue(jsonObjectCallback(callback))
    }

    fun updateObject(obj: ObjectItem, boxId: Long, callback: (Result<ObjectItem>) -> Unit) {
        val payload = mapOf("nombre" to obj.nombre, "state" to obj.state)
        val requestBody = json.encodeToString(payload).toRequestBody("application/json".toMediaType())
        val request = Request.Builder()
            .url("$baseURL/boxes/$boxId/objects/${obj.id}")
            .put(requestBody)
            .build()
        client.newCall(request).enqueue(jsonObjectCallback(callback))
    }

    fun deleteObject(id: Long, boxId: Long, callback: (Result<Unit>) -> Unit) {
        val request = Request.Builder()
            .url("$baseURL/boxes/$boxId/objects/$id")
            .delete()
            .build()
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                callback(Result.failure(APIError.InvalidResponse))
            }
            override fun onResponse(call: Call, response: Response) {
                if (!response.isSuccessful) {
                    callback(Result.failure(APIError.ServerError(response.code)))
                    return
                }
                callback(Result.success(Unit))
            }
        })
    }

    private fun jsonBoxCallback(callback: (Result<Box>) -> Unit): Callback = object : Callback {
        override fun onFailure(call: Call, e: IOException) {
            callback(Result.failure(APIError.InvalidResponse))
        }
        override fun onResponse(call: Call, response: Response) {
            if (!response.isSuccessful) {
                callback(Result.failure(APIError.ServerError(response.code)))
                return
            }
            val body = response.body?.string() ?: return callback(Result.failure(APIError.InvalidResponse))
            try {
                val dto = json.decodeFromString<BoxDTO>(body)
                callback(Result.success(dto.toBox()))
            } catch (e: Exception) {
                callback(Result.failure(APIError.DecodingError))
            }
        }
    }

    private fun jsonObjectCallback(callback: (Result<ObjectItem>) -> Unit): Callback = object : Callback {
        override fun onFailure(call: Call, e: IOException) {
            callback(Result.failure(APIError.InvalidResponse))
        }
        override fun onResponse(call: Call, response: Response) {
            if (!response.isSuccessful) {
                callback(Result.failure(APIError.ServerError(response.code)))
                return
            }
            val body = response.body?.string() ?: return callback(Result.failure(APIError.InvalidResponse))
            try {
                val dto = json.decodeFromString<ObjectItemDTO>(body)
                callback(Result.success(dto.toObject()))
            } catch (e: Exception) {
                callback(Result.failure(APIError.DecodingError))
            }
        }
    }
}